appmap/_implementation/generation.py

```diff
@@ -0,0 +1,117 @@
+"""Generate an AppMap"""
+import json
+from collections.abc import MutableMapping
+
+from . import metadata
+
+from .event import Event, serialize_event
+
+
+class ClassMapSet(MutableMapping):
+    def __init__(self):
+        self.dict = dict()
+
+    def __delitem__(self, k):
+        del self[k]
+
+    def __getitem__(self, k):
+        return self.dict[k]
+
+    def __setitem__(self, k, v):
+        self.dict[k] = v
+
+    def __iter__(self):
+        return self.dict.__iter__()
+
+    def __len__(self):
+        return len(self.dict)
+
+
+class ClassMapEntry:
+    def __init__(self, key, name, entry_type):
+        self.key = key
+        self.name = name
+        self.type = entry_type
+
+    def __eq__(self, other):
+        return self.key == other.key
+
+    def __hash__(self):
+        return hash(self.key)
+
+
+class PackageEntry(ClassMapEntry):
+    def __init__(self, name):
+        super().__init__(name, name, 'package')
+        self.children = ClassMapSet()
+
+    __hash__ = ClassMapEntry.__hash__
+
+
+class ClassEntry(ClassMapEntry):
+    def __init__(self, name):
+        super().__init__(name, name, 'class')
+        self.children = ClassMapSet()
+
+    __hash__ = ClassMapEntry.__hash__
+
+
+class FuncEntry(ClassMapEntry):
+    def __init__(self, e, loc):
+        super().__init__(loc, e.method_id, 'function')
+        self.path = e.path
+        self.lineno = e.lineno
+        self.static = e.static
+
+    __hash__ = ClassMapEntry.__hash__
+
+
+def asdict(s):
```

Comment:
@apotterri why a separate function?

Reply:
This was a holdover from an implementation using `orjson`, which required a separate function. I've moved this into `AppMapEncoder`.